tsl/src/continuous_aggs/invalidation_record.c

```diff
@@ -0,0 +1,120 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <funcapi.h>
+
+#include "invalidation_record.h"
+
+#include <access/tupdesc.h>
+#include <lib/stringinfo.h>
+#include <libpq/pqformat.h>
+#include <replication/logicalproto.h>
+
+#include "compat/compat.h"
+#include "export.h"
+#include <utils/elog.h>
+
+TS_FUNCTION_INFO_V1(ts_invalidation_read_record);
+
+/*
+ * Helper macros always use the current version when encoding the entry.
+ */
+#define _field(v) CppConcat(ver, v)
+#define _ver _field(INVALIDATION_ENTRY_CURRENT_VERSION)
+
+/*
+ * Read an encoded invalidation record coming from the plugin.
+ *
+ * Records from the plugin is coming in logical replication format, so decode
+ * it and produce a single composite value in the same format as the
+ * invalidation table.
+ */
+Datum
+ts_invalidation_read_record(PG_FUNCTION_ARGS)
+{
+	bytea *raw_record = PG_GETARG_BYTEA_P(0);
+
+	TupleDesc tupdesc;
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function returning record called in context "
+						"that cannot accept type record")));
+	tupdesc = BlessTupleDesc(tupdesc);
+
+	StringInfoData info;
+	initReadOnlyStringInfo(&info, VARDATA_ANY(raw_record), VARSIZE_ANY_EXHDR(raw_record));
+
+	InvalidationEntry entry;
+	ts_invalidation_record_decode(&info, &entry);
+
+	Datum values[] = {
+		[0] = ObjectIdGetDatum(entry._ver.hypertable_relid),
+		[1] = Int64GetDatum(entry._ver.lowest_modified),
+		[2] = Int64GetDatum(entry._ver.highest_modified),
+	};
+	bool nulls[3] = { 0 };
+
+	Assert(tupdesc->natts == sizeof(values) / sizeof(*values));
+
+	HeapTuple htup = heap_form_tuple(tupdesc, values, nulls);
+
+	PG_RETURN_DATUM(HeapTupleGetDatum(htup));
+}
+
+/*
+ * Decode invalidation record in byte array format.
+ *
+ * This need to handle all supported version formats, not just the current
+ * one. This function should be used on the receiving side of the invalidation
+ * plugin and be used to decode records arriving from the plugin.
+ */
+void
+ts_invalidation_record_decode(StringInfo record, InvalidationEntry *entry)
+{
+	/*
+	 * The header is the same for all versions of the invalidation record.
+	 */
+	entry->version = pq_getmsgint32(record);
```

Comment:
I don't think the version is needed in the struct, since we never look at it. It is enough to just check the version here.

Reply:
I expect this to be needed for validations in the future, so want to keep it around.